### ISSUE TITLE
About training code

### ISSUE THREAD

**Issue by IsabelleCur** [ShahinSHH/COVID-FACT#1]
Thanks a lot for sharing code! I've read the papers about the dataset and model. But I'm still confused about how to use the unlabeled subset to develop the second stage of the model. May I get the training code? Thanks again!
yingpeng2000@gmail.com

**Comment by ShahinSHH** [comment]
Thanks for your interest in the work. I can't upload the training code at the moment but I can provide you the way I implemented that. In brief, first, you can train the first stage (the codes of the model are provided in the repo, you just need to fit them with appropriate labels), then apply the stage 1 model on each patient and save the  identified infected slices, then use saved abnormal(infected) slices to train the second stage.
You can also train both stages simultaneously but it's not the way used in the paper.

**Comment by IsabelleCur** [comment]
OK, I got it. Thanks for in-time response!

**Comment by HoudaJebbari** [comment]
Hi, I also have a question regarding the training code. For the training of the first stage, I selected the identified infected slices of COVID and CAP cases and saved them in one directory and the other slices of COVID and CAP showing no infection as well as normal cases are saved in another directory. I separated the data (60, 30, 10) and generated the 3 CSV files for training, validation and testing and then copied all the images to one directory as they are already separated in the CSV files.

But I had many problems fiting the data. Can you tell me how you did it? It will be a great help
Thank you in advance.

**Comment by IsabelleCur** [comment]
I didn't train the data because I used the same dataset. I just input the testing data to get a result and computed the accuracy.

**Comment by HoudaJebbari** [comment]
Thank you for your answer. I am re-training the model because I am pre-processing the data with a filter.

**Comment by ShahinSHH** [comment]
> Hi, I also have a question regarding the training code. For the training of the first stage, I selected the identified infected slices of COVID and CAP cases and saved them in one directory and the other slices of COVID and CAP showing no infection as well as normal cases are saved in another directory. I separated the data (60, 30, 10) and generated the 3 CSV files for training, validation and testing and then copied all the images to one directory as they are already separated in the CSV files.
> 
> But I had many problems fiting the data. Can you tell me how you did it? It will be a great help
> Thank you in advance.

Hello, to train the second stage you can save the slices identified as infectious for COVID-19, CAP, and Normal cases separately, and then import them at the beginning of your 2nd stage training code. Please note that you should follow the same train/test split as you did for the first stage. I concatenate the identified slices for COVID-19 cases and labeled them as class 1, and combined all identified slices from CAP and Normal classes and labeled them as 0. Usually, you won't have many slices from the normal cases if you train the first stage well. I didn't use a CSV file to save train/valid/test splits but you can proceed with your approach for sure. Here is the part of code I used to fit the 2nd stage model:


adam = optimizers.Adam(lr=1e-4) 
model.compile(loss='binary_crossentropy', optimizer=adam, metrics=['accuracy'])
data_augmentation = False

batch_size = 16
epochs = 100
save_weight_path = 'your_save_dir/'
save_weight_name = 'stage2_lung_v1.h5' 
filepath = save_weight_path + save_weight_name

checkpoint = ModelCheckpoint(filepath, monitor='val_loss', verbose=1, save_best_only=True, mode='min')
callbacks_list = [checkpoint]

history = model.fit(x_train, y_train, batch_size = batch_size,
      epochs = epochs,
      validation_data = (x_valid, y_valid),
      shuffle = False,
      class_weight = class_weights,
      callbacks = callbacks_list)

**Comment by HoudaJebbari** [comment]
Ok, thanks for your answer. Is "_x_train_" the array returned by the "_segment_lung_" function? If so, and how did you label the elements of the array to create  "_y_train_" ?

**Comment by ShahinSHH** [comment]
> Ok, thanks for your answer. Is "_x_train_" the array returned by the "_segment_lung_" function? If so, and how did you label the elements of the array to create "_y_train_" ?

Let me clarify how you can create x_train and y_train for each stage. You may know parts of them but I think it would be good to clarify it here.
For stage 1, x_train is the output of the lung segmentation function. You can save the lungs and then concatenate them based on your train/test split. I used this dataset ([COVID-CT-MD](https://github.com/ShahinSHH/COVID-CT-MD)) to train and evaluate the model. In this dataset, slices with the evidence of infection (caused by CAP or COVID-19) are indicated. So you can start importing/preprocessing the data for each case and at the same time update your y_train using the labels (in a CSV file) provided in that dataset.
For stage 2:
to create the x_train, first, you need to feed the cases into the trained version of the stage 1 model and save the slices identified as infectious for each case in your device (you need to repeat this step for train/valid/test cases). Then, you should concatenate them and update your y_train. To create the y_train, when you import slices, if they belong to a COVID-19 case, you'll have to label them as class 1. Otherwise, they should be labeled as class 0.
